src/Microsoft.AspNet.OData/Query/PlainTextODataQueryOptionsParser.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+using Microsoft.OData;
+using Microsoft.AspNet.OData.Common;
+using System.Net.Http;
+using System.Net.Http.Headers;
+
+namespace Microsoft.AspNet.OData.Query
+{
+    /// <summary>
+    /// Reads and parses the content of a <see cref="T:System.IO.Stream" /> 
+    /// into a query options part of an OData URL. 
+    /// The query options are passed in the request body as plain text.
+    /// </summary>
+    /// <remarks>This class derives from a platform-specific class.</remarks>
+    [SuppressMessage("Microsoft.Naming", "CA1702:CompoundWordsShouldBeCasedCorrectly", Justification = "Spelling correct in this context")]
+    public partial class PlainTextODataQueryOptionsParser : IODataQueryOptionsParser
+    {
+        private static MediaTypeHeaderValue supportedMediaType = MediaTypeHeaderValue.Parse("text/plain");
+
+        /// <inheritdoc/>
+        public bool CanParse(HttpRequestMessage request)
+        {
+            if (request == null)
+            {
+                throw Error.ArgumentNull("httpRequest");
+            }
+
+            MediaTypeHeaderValue contentType = request.Content.Headers.ContentType;
+
+            return contentType?.MediaType?.StartsWith(supportedMediaType.MediaType, StringComparison.Ordinal) == true ? true : false;
+        }
+
+        /// <inheritdoc/>
+        public async Task<string> ParseAsync(Stream requestStream)
+        {
+            string queryString = string.Empty;
+
+            MemoryStream memoryStream = new MemoryStream();
+            TextReader reader;
+
+            try
+            {
+                await requestStream.CopyToAsync(memoryStream);
+                memoryStream.Position = 0;
+                reader = new StreamReader(memoryStream);
+            }
+            catch
+            {
+                memoryStream.Dispose();
+
+                throw new ODataException(SRResources.CannotParseQueryOptionsPayload);
+            }
```

Comment:
What is the value of copying to a memory stream, versus reading directly from the requestStream?

Reply:
@mikepizzo The idea of copying to a memory stream was to avoid tinkering with the request stream - at the time I thought that `CopyTo`/`CopyToAsync` do not progress the stream position. I just did some tests that confirmed that thinking to be wrong so I'll change the logic to just read directly from the request stream